### ISSUE TITLE
PolyStrips on thin surfaces can cause crashes that never end

### ISSUE THREAD

**Issue by vxlcoder** [CGCookie/retopoflow#217]
When drawing polystrips across thin models it frequently crashes and stalls Blender.

**Comment by carter2422** [comment]
Were you able to replicate this @vxlcoder?

**Comment by vxlcoder** [comment]
i was.  there was a bug in the code that tests if `bvh.ray_cast` returns `None` instead of checking if it returns a tuple of `None`s.  see https://www.blender.org/api/blender_python_api_2_75_3/mathutils.bvhtree.html#mathutils.bvhtree.BVHTree.ray_cast

**Comment by carter2422** [comment]
Ah excellent! What were the steps to reproduce it? I'd like to verify as well so as to let the user know.

**Comment by vxlcoder** [comment]
running polystrips on a very thin surface (cylinder without caps, scaled in the faces extruded by just a bit) with crossing strokes seemed to reproduce the error.  it had to do with snapping the verts on a thin surface.  actually, blender 2.76a would segfault around here; 2.76b fixed the issue.

**Comment by carter2422** [comment]
Suffice to say I could not replicate this:
![screen shot 2016-03-07 at 9 38 32 pm](https://cloud.githubusercontent.com/assets/1648631/13591315/fa54aa42-e4ac-11e5-80a0-86f31ca9cabd.png)

Neither in 2.76b or 2.77rc1.

**Comment by carter2422** [comment]
Ah ha! I found a way to replicate this @vxlcoder. 

The issue is when a strip is drawn *across* a thin model, such that it's normal oppose each other from one side of the strip to the other. 

See these examples:
![Cylinder scaled along the X, scale applied](https://cloud.githubusercontent.com/assets/1648631/13882526/c98e6f8e-ecf3-11e5-98cd-4c41456be0f4.png)

![Viewed from the front](https://cloud.githubusercontent.com/assets/1648631/13882550/e92af632-ecf3-11e5-9afa-ab4e4b302afb.png)

![Draw strip left to right, crossing both sides](https://cloud.githubusercontent.com/assets/1648631/13882551/ecb4c364-ecf3-11e5-9a84-b7ccb6486a61.png)

Instant crash:
```
EXCEPTION (<class 'AttributeError'>): 'NoneType' object has no attribute 'dot'
         /Users/jonathanwilliamson/Dropbox (Personal)/blenderDev/addons/cgCookie/scripts/addons/retopoflow/op_polystrips/polystrips_datastructure.py
000 1251:snap_igverts() d0,d3 = nr.dot(self.gvert0.snap_norm),nr.dot(self.gvert3.snap_norm)
001 1153:update_nozip() self.snap_igverts()
002 1176:update() self.update_nozip(debug=debug)
003 0305:update() if gedge: gedge.update()
004 0216:update_gedges() self.update()
005 0228:connect_gedge() self.update_gedges()
006 0600:__init__() gvert0.connect_gedge(self)
007 2077:create_gedge() self.length_scale, gv0, gv1, gv2, gv3)
008 2581:insert_gedge_from_stroke() self.create_gedge(gv0,gv1,gv2,gv3)
009 2333:insert_gedge_from_stroke() self.insert_gedge_from_stroke(stroke[:i00], only_ends, sgv0=sgv0, sgv3=gv_intersect, depth=depth+1)
         /Users/jonathanwilliamson/Dropbox (Personal)/blenderDev/addons/cgCookie/scripts/addons/retopoflow/op_polystrips/polystrips_ui_tools.py
010 0111:modal_sketching() self.polystrips.insert_gedge_from_stroke(p3d, False)
         /Users/jonathanwilliamson/Dropbox (Personal)/blenderDev/addons/cgCookie/scripts/addons/retopoflow/modaloperator.py
011 0276:modal() nmode = self.FSM[self.fsm_mode](context, eventd)
```

**Comment by carter2422** [comment]
It seems your fixes in 993a0cbfbac1de13d24a9d9b0980ccd9ff3e502d did the trick @vxlcoder. I'm going to merge into release.
